tests/vm_console_proxy_test.go

```diff
@@ -0,0 +1,177 @@
+	var (
+		clusterRoleResource        testResource
+		clusterRoleBindingResource testResource
+		serviceAccountResource     testResource
+		serviceResource            testResource
+		deploymentResource         testResource
+		configMapResource          testResource
+	)
+
+	BeforeEach(func() {
+		expectedLabels := expectedLabelsFor("vm-console-proxy", "vm-console-proxy")
+		clusterRoleResource = testResource{
+			Name:           "vm-console-proxy",
+			Resource:       &rbac.ClusterRole{},
+			ExpectedLabels: expectedLabels,
+			UpdateFunc: func(role *rbac.ClusterRole) {
+				role.Rules[0].Verbs = []string{"watch"}
+			},
+			EqualsFunc: func(old *rbac.ClusterRole, new *rbac.ClusterRole) bool {
+				return reflect.DeepEqual(old.Rules, new.Rules)
+			},
+		}
+		clusterRoleBindingResource = testResource{
+			Name:           "vm-console-proxy",
+			Resource:       &rbac.ClusterRoleBinding{},
+			ExpectedLabels: expectedLabels,
+			UpdateFunc: func(roleBinding *rbac.ClusterRoleBinding) {
+				roleBinding.Subjects = nil
+			},
+			EqualsFunc: func(old *rbac.ClusterRoleBinding, new *rbac.ClusterRoleBinding) bool {
+				return reflect.DeepEqual(old.RoleRef, new.RoleRef) &&
+					reflect.DeepEqual(old.Subjects, new.Subjects)
+			},
+		}
+		serviceAccountResource = testResource{
+			Name:           "vm-console-proxy",
+			Namespace:      strategy.GetVmConsoleProxyNamespace(),
+			Resource:       &core.ServiceAccount{},
+			ExpectedLabels: expectedLabels,
+			UpdateFunc: func(serviceAccount *core.ServiceAccount) {
+				serviceAccount.Secrets = nil
+			},
+			EqualsFunc: func(old *core.ServiceAccount, new *core.ServiceAccount) bool {
+				return reflect.DeepEqual(old.Secrets, new.Secrets)
+			},
+		}
+		serviceResource = testResource{
+			Name:           "vm-console-proxy",
+			Namespace:      strategy.GetVmConsoleProxyNamespace(),
+			Resource:       &core.Service{},
+			ExpectedLabels: expectedLabels,
+			UpdateFunc: func(service *core.Service) {
+				service.Spec.Ports[0].Port = 1443
+				service.Spec.Ports[0].TargetPort = intstr.FromInt(18768)
+			},
+			EqualsFunc: func(old *core.Service, new *core.Service) bool {
+				return reflect.DeepEqual(old.Spec, new.Spec)
+			},
+		}
+		deploymentResource = testResource{
+			Name:           "vm-console-proxy",
+			Namespace:      strategy.GetVmConsoleProxyNamespace(),
+			Resource:       &apps.Deployment{},
+			ExpectedLabels: expectedLabels,
+			UpdateFunc: func(deployment *apps.Deployment) {
+				deployment.Spec.Replicas = pointer.Int32Ptr(0)
+			},
+			EqualsFunc: func(old *apps.Deployment, new *apps.Deployment) bool {
+				return reflect.DeepEqual(old.Spec, new.Spec)
+			},
+		}
+		configMapResource = testResource{
+			Name:           "vm-console-proxy",
+			Namespace:      strategy.GetVmConsoleProxyNamespace(),
+			Resource:       &core.ConfigMap{},
+			ExpectedLabels: expectedLabels,
+			UpdateFunc: func(configMap *core.ConfigMap) {
+				configMap.Data = nil
```

Comment:
It's better to add some unique data, because it may already be `nil` and we want to test an update.

Reply:
Ack.